pf/internal/defaults.go

```diff
@@ -1,42 +1,139 @@
+		raw, err := defaultInfo.From(res)
+		if err != nil {
+			return na, false, err
+		}
+		if raw == nil {
+			return na, false, nil
+		}
+		return recoverDefaultValue(raw), true, nil
+
+	} else if defaultInfo.EnvVars != nil {
+		for _, n := range defaultInfo.EnvVars {
+			if str, ok := os.LookupEnv(n); ok {
+				// TODO what is the behavior of an env var set to an emtpy string? Is this the same as
+				// unset, or the same as actually setting to empty?
+				v, err := parseValueFromEnv(fieldSchema, str)
+				return v, true, err
 			}
 		}
 	}
-	return d.Value, nil
+	if defaultInfo.Value != nil {
+		return recoverDefaultValue(defaultInfo.Value), true, nil
+	}
+	return na, false, nil
 }
 
-func SetDefaultValues(res *tfbridge.PulumiResource, fields map[string]*tfbridge.SchemaInfo) error {
-	for key, fld := range fields {
-		if _, ok := res.Properties[resource.PropertyKey(key)]; !ok && fld.Default != nil {
-			// using default value for empty property
-			v, err := getDefaultValue(fld.Default, res)
-			if err != nil {
-				return err
+func parseValueFromEnv(sch shim.Schema, str string) (resource.PropertyValue, error) {
+	var err error
+	switch sch.Type() {
+	case shim.TypeBool:
+		v := false
+		if str != "" {
+			if v, err = strconv.ParseBool(str); err != nil {
+				return resource.NewNullProperty(), err
+			}
+		}
+		return resource.NewBoolProperty(v), nil
+	case shim.TypeInt:
+		v := int(0)
+		if str != "" {
+			iv, iverr := strconv.ParseInt(str, 0, 0)
+			if iverr != nil {
+				return resource.NewNullProperty(), iverr
 			}
-			if pval, ok := v.(resource.PropertyValue); ok {
-				res.Properties[resource.PropertyKey(key)] = pval
+			v = int(iv)
+		}
+		return resource.NewNumberProperty(float64(v)), nil
+	case shim.TypeFloat:
+		v := float64(0.0)
+		if str != "" {
+			if v, err = strconv.ParseFloat(str, 64); err != nil {
+				return resource.NewNullProperty(), err
 			}
-			res.Properties[resource.PropertyKey(key)] = resource.NewPropertyValue(v)
 		}
+		return resource.NewNumberProperty(float64(v)), nil
+	case shim.TypeString:
+		return resource.NewStringProperty(str), nil
+	default:
+		return resource.NewNullProperty(), fmt.Errorf("unknown type for default value: %v", sch.Type())
+	}
+}
+
+func recoverDefaultValue(defaultValue any) resource.PropertyValue {
+	if pv, alreadyPV := defaultValue.(resource.PropertyValue); alreadyPV {
+		return pv
 	}
+	return resource.NewPropertyValue(defaultValue)
+}
+
+// TODO this needs to handle the nested case, setting defaults for properties of object-typed properties.
+func SetDefaultValues(
+	res *tfbridge.PulumiResource,
+	resShim shim.Resource,
+	fields map[string]*tfbridge.SchemaInfo,
+) (finalError error) {
+	for key, fld := range fields {
+		if fld == nil || fld.Default == nil {
+			continue
+		}
+		fieldSchema, knownField := resShim.Schema().GetOk(key)
+		if !knownField {
+			continue
+		}
 
-	return nil
+		pulumiName := tfbridge.TerraformToPulumiNameV2(key, resShim.Schema(), fields)
```

Comment:
@tmeckel this name translation bit is unfortunately very important, otherwise foo_bar and fooBar names are confused and there's map mismatches.

Reply:
@t0yv0 shall we get in sync with the naming of tfbridge code and use `applyDefaults` as the function name?

Reply:
applyDefaults does a lot more , this one applies only bridge-specified defaults from DefaultInfo; the other one tries to implement TF behavior of applying TF-specified defaults. So a new name is warranted but naming is hard.